File: c4_linkedlist/L61-m.cpp
//
// Created by LiHeng on 2020/4/19.
//
#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == NULL || head->next == NULL) return head;
        int len = 0;
        ListNode* cnt = head;
        ListNode* tail = head; //先把尾节点标记下来
        while(cnt){
            ++len;
            tail = cnt;
            cnt = cnt->next;
        }
        if(k%len == 0) return head;
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* pre = dummy;
        ListNode* cur = head;
        for(int i=0; i<(len-k%len); ++i){
            cur = cur->next;
            pre = pre->next;
        }
        dummy->next = cur;
        tail->next = head;
        pre->next = NULL;
        return dummy->next;
    }
};

//也可以下面这么写，思路是一样的
class Solution1 {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL)
            return head;
        int len = 1;
        ListNode* tail = head;
        while(tail->next != NULL){
            tail = tail->next;
            ++len;
        }
        if(k%len == 0)
            return head;

        ListNode* last = head;
        for(int i=0 ; i < len-k%len-1; i++)
            last = last->next;

        ListNode* new_start = last->next;
        last->next = NULL;
        tail->next = head;

        return new_start;
    }

};

File: python-algorithm/ds_4_linkedlist/linkedlist_2_inplace_operation/L61-m.py
#--coding:utf-8--

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        cnt = 1
        tail = head
        # 计算出链表长度并标记最后个节点
        while tail.next:
            cnt += 1
            tail = tail.next
        if k%cnt == 0:
            return head
        slow = fast = head
        for _ in range(k%cnt):
            fast = fast.next
        while fast.next:
            slow = slow.next
            fast = fast.next
        new_head = slow.next
        slow.next = None
        tail.next = head
        return new_head
